### DISCUSSION TITLE
Error using mongoose from a library

### DISCUSSION THREAD

**Discussion by softhardwareguy** [cesanta/mongoose#2708] [Ask questions, get answers] [answered]
I am getting a rather odd error when trying to build from a mongoose library. Basically I built mongoose into a library (using the mongoose make file, target linux_libs or something like that) and I am going to link to it. I do this because I need to keep our OSS out of our repo and separate for legal reasons. When I do this, I get the below build issues. You will note it is finding mongoose.h (not throwing an error not finding the header file), but not anything defined in it. But if I put mongoose.h in the directory next to main.c, then it builds main.c just fine. What am I missing here? My read of the header say I don't need to pass anything to the header for it to build correctly on linux x86/64. 

Thanks,
Tom

/u/gcc-13.2.0-linux_x86_64/bin/gcc -D_FILE_OFFSET_BITS=64 -MMD -MP -g -gdwarf-4  -m64 -fPIC -O3  -g -gdwarf-4  -m64 -fPIC -O3  -DNGU_BETA=1 -DNGU_DEBUG=0 -DNGU_GE_DEBUG=0 -DNGU_VIS_DEBUG=0  -DNGU_BETA=1 -Wall -Wredundant-decls -Wpacked -Wmissing-noreturn -Wsign-compare -Wconversion -Wundef -Wshadow -funit-at-a-time -Wtype-limits -I/u/oss/valijson/1.0/include -isystem /u/oss/boost/1.78/gcc-11.2.0/include -isystem /u/oss/hocon/0.3.0/gcc-11.2.0/include -isystem /u/oss/leatherman/1.4.0/gcc-11.2.0/include -isystem /u/oss/icu/50.2/gcc-11.2.0/include -fcompare-debug-second -I/u/oss/mongoose/7.12/gcc-13.2.0/usr/local/include -I/u/oss/cJSON/1.7.17/gcc-5.3.0/include -I/u/NextGenUI/src/include -I/u/NextGenUI/src/gui -I/u/NextGenUI/src/server -I/u/NextGenUI/build/tmp/linux_x86_64 -I/u/NextGenUI/build/tmp/linux_x86_64  -c /u/NextGenUI/src/server/main.c -o /u/NextGenUI/build/tmp/linux_x86_64/main.o
/u/NextGenUI/src/server/main.c: In function ‘thread_function’:
/u/NextGenUI/src/server/main.c:30:3: warning: implicit declaration of function ‘mg_wakeup’ [-Wimplicit-function-declaration]
   30 |   mg_wakeup(p->mgr, p->conn_id, "hi!", 3);  // Respond to parent
      |   ^~~~~~~~~
/u/NextGenUI/src/server/main.c: In function ‘fn’:
/u/NextGenUI/src/server/main.c:52:20: error: ‘MG_EV_WAKEUP’ undeclared (first use in this function)
   52 |   } else if (ev == MG_EV_WAKEUP) {
      |                    ^~~~~~~~~~~~
/u/NextGenUI/src/server/main.c: In function ‘main’:
/u/NextGenUI/src/server/main.c:62:49: warning: passing argument 3 of ‘mg_http_listen’ from incompatible pointer type [-Wincompatible-pointer-types]
   62 |   mg_http_listen(&mgr, "http://localhost:8000", fn, NULL);  // Create listener
      |                                                 ^~
      |                                                 |
      |                                                 void (*)(struct mg_connection *, int,  void *)
/u/NextGenUI/src/server/main.c:63:3: warning: implicit declaration of function ‘mg_wakeup_init’; did you mean ‘mg_queue_init’? [-Wimplicit-function-declaration]
   63 |   mg_wakeup_init(&mgr);  // Initialise wakeup socket pair
      |   ^~~~~~~~~~~~~~
      |   mg_queue_init

**Comment by scaprile** [comment, marked as answer]
We have an extensive documentation, and a rather huge set of examples.
All the examples are built on every push, and on every night we build on LInux, Mac, Windows, for Arm Cortex-M, ESP32.
It is impossible for me to know what you are doing or not doing if I don't see what you are doing (or not doing). Besides that, we don't check user code.
We are talking here of 2 files: a c file that has to be on your compiler path (now a library file on your linker path), and a header file that has to be on your include path.
The only thing that I can possibly come up with is that, for some reason, your architecture is not being detected and some functions not included. If that is the case, you should define the architecture or the sockets interface https://mongoose.ws/documentation/#build-options
If there is an actual issue, please open an issue, _honor the issue template_ so we can know your scenario and reproduce the issue if necessary, and let us know _exactly_ what you are doing
If you are a paying customer, please contact the proper email as per your contract. Thank you.

Please follow the guidelines in the [documentation](https://mongoose.ws/documentation/),  [examples](https://github.com/cesanta/mongoose/tree/master/examples/) and [tutorials](https://mongoose.ws/documentation/#tutorials-and-examples) available.

**Reply by softhardwareguy** [reply]
This is one of the examples. main.c is actually the multi-threaded example. I can run the example using the simplistic place the code next to the example code methodology, but as I stated above, approaching it in a more organized fashion of pre-building the mongoose library and it is failing in an odd way I have never seen before.

**Reply by scaprile** [reply]
Comment updated, please re-read
